rust/agents/scraper/src/agent.rs

```diff
@@ -106,69 +115,127 @@ impl BaseAgent for Scraper {
-            {
-                trace!(%domain, "Created mailbox and indexer");
-                let scraper = SqlChainScraper::new(
+            let db = Arc::new(
+                HyperlaneSqlDb::new(
                     db.clone(),
-                    local,
-                    &chain_setup.index,
-                    contract_sync_metrics.clone(),
+                    chain_setup.addresses.mailbox,
+                    domain.clone(),
+                    settings
+                        .build_provider(domain, &metrics.clone())
+                        .await?
+                        .into(),
+                    &chain_setup.index.clone(),
                 )
-                .await?;
-                scrapers.insert(domain.id(), scraper);
-            }
+                .await?,
+            );
+            scrapers.insert(
+                domain.id(),
+                ChainScraper {
+                    domain: domain.clone(),
+                    db,
+                    index_settings: chain_setup.index.clone(),
+                },
+            );
         }
 
-        trace!(domain_count = scrapers.len(), "Creating scraper");
+        trace!(domain_count = scrapers.len(), "Created scrapers");
 
-        Ok(Self { core, db, scrapers })
+        Ok(Self {
+            core,
+            metrics,
+            contract_sync_metrics,
+            scrapers,
+        })
     }
 
     #[allow(clippy::async_yields_async)]
     async fn run(&self) -> Instrumented<JoinHandle<eyre::Result<()>>> {
-        let tasks = self
-            .scrapers
-            .iter()
-            .map(|(name, scraper)| {
-                let span = info_span!("ChainContractSync", %name, chain=%scraper.domain());
-                tokio::spawn(scraper.clone().sync()).instrument(span)
-            })
-            .collect();
-
+        let mut tasks = Vec::with_capacity(self.scrapers.len());
+        for domain in self.scrapers.keys() {
+            tasks.push(self.scrape(domain).await);
+        }
         run_all(tasks)
     }
 }
 
 impl Scraper {
-    async fn load_chain(
-        config: &Settings,
-        domain: &HyperlaneDomain,
-        metrics: &Arc<CoreMetrics>,
-    ) -> eyre::Result<Contracts> {
-        macro_rules! b {
-            ($builder:ident) => {
-                config
-                    .$builder(domain, metrics)
+    /// Sync contract data and other blockchain with the current chain state.
+    /// This will create a long-running task that should be spawned.
+    async fn scrape(&self, domain_id: &u32) -> Instrumented<JoinHandle<eyre::Result<()>>> {
+        let scraper = self.scrapers.get(domain_id).unwrap();
+        let db = scraper.db.clone();
+        let index_settings = scraper.index_settings.clone();
+        let domain = scraper.domain.clone();
+
+        let mut tasks = Vec::with_capacity(2);
+        let message_sync = self
+            .as_ref()
+            .settings
+            .build_message_sync(
```

Comment:
That builder macro I had reduces a lot of this boilerplate.

Reply:
Created a similar macro